src/tests/test_csv_parser.cpp

```diff
@@ -0,0 +1,95 @@
+                            std::vector<std::vector<std::string>> const& expected) {
+    auto parser = table.MakeInputTable();
+    if (table.has_header) num_rows--;
+
+    std::vector<std::vector<std::string>> actual;
+    for (size_t index = 0; index < num_rows; index++) {
+        std::vector<std::string> row = parser->GetNextRow();
+        actual.push_back(std::move(row));
+    }
+
+    ASSERT_THAT(actual, ContainerEq(expected)) << "Fail on " << table.name;
+}
+
+TEST(TestCSVParser, TestGetNextRow) {
+    CheckGetNextRow(kNullEmpty, 4,
+                    {{"1", "NULL", "3", "1"}, {"1", "2", "", "1"}, {"1", "2", "3", "1"}});
+    CheckGetNextRow(kTestSingleColumn, 5, {{"1"}, {"2"}, {"3"}, {"3"}});
+    CheckGetNextRow(kTestWide, 3, {{"1", "3", "3", "4", "5"}, {"2", "3", "4", "4", "6"}});
+    CheckGetNextRow(kTestEmpty, 1, {});
+    CheckGetNextRow(kabalone, 1,
+                    {{"M", "0.455", "0.365", "0.095", "0.514", "0.2245", "0.101", "0.15", "15"}});
+    CheckGetNextRow(kTestParse, 5,
+                    {{"", "\\\\\\\"", "b\"b\\\\ b"},
+                     {"\"", "\\\\", "b\\"},
+                     {"a,bc", "a,\"bc", "a\",bc"},
+                     {"bb", "\\\\", "\\\\"},
+                     {"a", "a,a", "a"}});
+}
+
+static void CheckHasNextRow(tests::TableConfig const& table, std::size_t num_rows) {
+    auto parser = table.MakeInputTable();
+    if (table.has_header) num_rows--;
+
+    for (size_t index = 0; index < num_rows; index++) {
+        bool has_next = parser->HasNextRow();
+        ASSERT_TRUE(has_next) << "Fail on " << table.name << ": expected true, actual false";
+        parser->GetNextRow();
+    }
+
+    bool has_next = parser->HasNextRow();
+    ASSERT_FALSE(has_next) << "Fail on " << table.name << ": expected false, actual true";
+}
+
+TEST(TestCSVParser, TestHasNextRow) {
+    CheckHasNextRow(kACShippingDates, 6);
+    CheckHasNextRow(kadult, 32561);
+    CheckHasNextRow(kTestEmpty, 1);
+    CheckHasNextRow(kTest1, 20);
+}
+
+static void CheckReset(tests::TableConfig const& table, std::size_t num_rows) {
+    auto parser = table.MakeInputTable();
+    if (table.has_header) num_rows--;
+
+    std::vector<std::vector<std::string>> first_parse;
+
+    for (size_t index = 0; index < num_rows; index++) {
+        std::vector<std::string> row = parser->GetNextRow();
+        first_parse.push_back(std::move(row));
+    }
+
+    parser->Reset();
+
+    std::vector<std::vector<std::string>> second_parse;
+    for (size_t index = 0; index < num_rows; index++) {
+        std::vector<std::string> row = parser->GetNextRow();
+        second_parse.push_back(std::move(row));
+    }
+
+    ASSERT_THAT(first_parse, ContainerEq(second_parse)) << "Fail on " << table.name;
+}
+
+TEST(TestCSVParser, TestReset) {
+    CheckReset(kACShippingDates, 6);
+    CheckReset(kadult, 32561);
+    CheckReset(kTestEmpty, 1);
+    CheckReset(kTest1, 20);
+}
+
+}  // namespace tests
```

Comment:
There must be line breaks at the end of all files (we also recently added #325 a line break at the end of files for csv tables, so it’s worth adding newline in `TestParse.csv`)